src/Groups/libraries/smallgroups.jl

```diff
@@ -1,89 +1,198 @@
+"""
+function small_group(::Type{T}, n::IntegerUnion, m::IntegerUnion) where T
+  G = _small_group(n, m)
+  return T(G)
+end
+
+function small_group(n::IntegerUnion, m::IntegerUnion)
+  G = _small_group(n, m)
   T = _get_type(G)
   return T(G)
 end
 
+function _small_group(n::IntegerUnion, m::IntegerUnion)
+  N = number_small_groups(n)
+  m <= N || throw(ArgumentError("There are only $N groups of order $n, up to isomorphism."))
+  return GAP.Globals.SmallGroup(GAP.Obj(n), GAP.Obj(m))
+end
+
+
 """
     small_group_identification(G::Group)
 
 Return `(n, m)`, where `G` is isomorphic with `small_group(n, m)`.
+
+# Examples
+```jldoctest
+julia> small_group_identification(alternating_group(5))
+(60, 5)
+```
 """
 function small_group_identification(G::GAPGroup)
-  r = GAP.Globals.IdGroup(G.X)
-  return (r[1], r[2])
+   isfinite(G) || error("group is not finite")
+   res = GAP.Globals.IdGroup(G.X)
+   res !== GAP.Globals.fail || error("identification is not available for groups of order $(order(G))")
+   return Tuple{Int,Int}(res)
 end
 
+
 """
-    number_small_groups(n::Int)
+    number_small_groups(n::IntegerUnion)
 
 Return the number of groups of order `n`, up to isomorphism.
 """
-number_small_groups(n::Int) = GAP.Globals.NumberSmallGroups(n)
+function number_small_groups(n::IntegerUnion)
+    n >= 1 || throw(ArgumentError("group order must be positive, not $n"))
+    return fmpz(GAP.Globals.NumberSmallGroups(n))
+end
 
 
 """
-    all_small_groups(n::Int, L...)
+    all_small_groups(L...)
 
-Return the list of all groups (up to isomorphism) of order `n` and satisfying
-the conditions in `L`. Here, `L` is a vector whose arguments are organized as
-`L` = [ `func1`, `arg1`, `func2`, `arg2`, ... ], and the function returns all
-the groups `G` satisfying the conditions `func1`(`G`) = `arg1`, `func2`(`G`) =
-`arg2`, etc. An argument can be omitted if it corresponds to the boolean value
-`true`.
+Return the list of all groups (up to isomorphism)
+satisfying the conditions describe by the arguments. These conditions
+may be of one of the following forms:
 
-The following command returns the list of all abelian non-cyclic groups
-of order 12.
+- `func => intval` selects groups for which the function `func` returns `intval`
+- `func => list` selects groups for which the function `func` returns any element inside `list`
+- `predicate` selects groups for which the function `predicate` returns `true`
+- `!predicate` selects groups for which the function `predicate` returns `false`
+
+As a special case, the first argument may also be one of the following:
+- `intval` selects groups whose order equals `intval`; this is equivalent to `order => intval`
+- `intlist` selects groups whose order is in `intlist`; this is equivalent to `order => intlist`
+
+Note that at least one of the conditions must impose a limit on the group
+order, otherwise an error is raised.
+
+The type of the returned groups is `PcGroup` if the group is solvable, `PermGroup` otherwise.
```

Comment:
Eventually (not in this pull request) we can support a type as the first argument of `all_small_groups`, with the same meaning as in `small_group`.